source/opt/decoration_manager.cpp

```diff
@@ -180,46 +235,42 @@ template <typename T>
 std::vector<T> DecorationManager::InternalGetDecorationsFor(
     uint32_t id, bool include_linkage) {
   std::vector<T> decorations;
-  std::stack<uint32_t> ids_to_process;
-
-  const auto process = [&ids_to_process, &decorations](T inst) {
-    if (inst->opcode() == SpvOpGroupDecorate ||
-        inst->opcode() == SpvOpGroupMemberDecorate)
-      ids_to_process.push(inst->GetSingleWordInOperand(0u));
-    else
-      decorations.push_back(inst);
-  };
 
   const auto ids_iter = id_to_decoration_insts_.find(id);
   // |id| has no decorations
   if (ids_iter == id_to_decoration_insts_.end()) return decorations;
 
-  // Process |id|'s decorations. Some of them might be groups, in which case
-  // add them to the stack.
-  for (ir::Instruction* inst : ids_iter->second) {
+  const TargetData& target_data = ids_iter->second;
+
+  // Process |id|'s decorations.
+  decorations.reserve(target_data.direct_decorations.size());
+  for (ir::Instruction* inst : ids_iter->second.direct_decorations) {
     const bool is_linkage =
         inst->opcode() == SpvOpDecorate &&
         inst->GetSingleWordInOperand(1u) == SpvDecorationLinkageAttributes;
-    if (include_linkage || !is_linkage) process(inst);
+    if (include_linkage || !is_linkage) decorations.push_back(inst);
   }
 
-  // If the stack is not empty, then it contains groups ID: retrieve their
-  // decorations and process them. If any of those decorations is applying a
-  // group, push that group ID onto the stack.
-  while (!ids_to_process.empty()) {
-    const uint32_t id_to_process = ids_to_process.top();
-    ids_to_process.pop();
-
-    // Retrieve the decorations of that group
-    const auto group_iter = group_to_decoration_insts_.find(id_to_process);
-    if (group_iter != group_to_decoration_insts_.end()) {
-      // Process all the decorations applied by the group.
-      for (T inst : group_iter->second) process(inst);
-    } else {
-      // Something went wrong.
-      assert(false);
-      return std::vector<T>();
-    }
+  // Process the decorations of all groups applied to |id|.
+  std::stack<uint32_t> groups_to_process;
```

Comment:
Just use an std::vector.  We want to avoid more uses of std::stack, std::queue, std::list, and std::deueque.  They perform very poorly on windows.

Reply:
Okay, I’ll do the change to std::vector.